src/fswatcher.c

```diff
@@ -0,0 +1,988 @@
+ * removed. NOTE: 0 is a special key in that it will be used in output for
+ * errors which were not directly the result of a command.
+ *
+ * When using a WATCH command with a <timestamp> argument, the timestamp must
+ * be an integer number of nanoseconds since Jan 1, 1970 00:00:00 UTC. This
+ * usually will be the time that the last event or modification was seen for
+ * this file and the data can be pulled from stat(2)'s st_mtim structure.
+ *
+ * On STDOUT you will see JSON messages that look like the following but are
+ * on a single line:
+ *
+ *  {
+ *     "changes": [array],
+ *     "code": <number>,
+ *     "is_final": true|false,
+ *     "key": <number>,
+ *     "message": "human readable string",
+ *     "pathname": "/path/which/had/event",
+ *     "result": "SUCCESS|FAIL",
+ *     "timestamp": <number>,
+ *     "type": <type>
+ *  }
+ *
+ * Where:
+ *
+ *   changes   is an array of strings indicating which changes occurred
+ *   code      is a positive integer code for an error
+ *   is_final  true when the event being printed is the last without re-watch
+ *   key       is the <KEY> for which a response corresponds
+ *   message   is a human-readable string describing response
+ *   pathname  is the path to which an event applies
+ *   result    indicates whether a command was a SUCCESS or FAILURE
+ *   timestamp timestamp in ns since 1970-01-01 00:00:00 UTC
+ *   type      is one of: event, response, error
+ *
+ * And:
+ *
+ *   "changes" is included only when (type == event)
+ *   "code" is included only when (type == error)
+ *   "is_final" is included when (type == event)
+ *   "key" is included whenever (type == response)
+ *   "message" is included whenever (type == error)
+ *   "pathname" is included whenever (type == event)
+ *   "result" is included when (type == response) value: "SUCCESS" or "FAILURE"
+ *   "timestamp" is included when (type == event)
+ *   "type" is always included
+ *
+ * Current values for "code" are in the ErrorCodes enum below.
+ *
+ * EXIT STATUS
+ *
+ *   Under normal operation, fswatcher will run until STDIN is closed or a fatal
+ *   error occurs. STDIN closing will result in exit code 0. Any other exit code
+ *   should result in a message of type "error" being output before exiting
+ *   non-zero.
+ *
+ *   When errors occur that are completly unexpected, this will call abort() to
+ *   generate a core dump.
+ *
+ */
+
+#define MAX_FMT_LEN 64
+#define MAX_KEY 4294967295
+#define MAX_KEY_LEN 10 /* number of digits (0-4294967295) */
+#define MAX_STAT_RETRY 10 /* number of times to retry stat() before abort() */
+#define MAX_TIMESTAMP_LEN 20
+
+/* longest command is '<KEY> UNWATCH <path>' with <KEY> being 10 characters. */
+#define MAX_CMD_LEN (MAX_KEY_LEN + 1 + 7 + 1 + PATH_MAX + 1 + MAX_TIMESTAMP_LEN)
+#define HANDLES_MASK 0xffff /* number of lookup buckets in the hash */
+#define SYSTEM_KEY 0
+
+/* hashing implementation, really needs a real one */
```

Comment:
Perhaps switch to something like djb2? (http://www.cse.yorku.ca/~oz/hash.html) To me that seems easier to follow than this macro. It's also not clear why we bundle strlen() functionality into this HASH macro. Is there a reason for this that I'm missing? The additional side-effect makes this even harder to follow.